source/link/linker.cpp

```diff
@@ -16,1183 +16,277 @@
+    if (res.second) {
+      ++i;
+      continue;
     }
-
-    if (version > max_version)
-      max_version = version;
-
-    auto &tmp_binary = tmp_binaries[i];
-    tmp_binary.resize(word_count);
-    std::memcpy(tmp_binary.data(), binary, word_count * sizeof(spirv_word));
-
-    uint32_t binary_max_id = GetWord(binary, 3u) - 1u;
-
-    spv_binary_t tmp_spvbinary = { tmp_binary.data(), word_count };
-    ret = TransformBinary2(hijack_context, tmp_spvbinary, BumpIDs);
-    if (ret != SPV_SUCCESS)
-      return ret;
-
-    id_upper_bound += binary_max_id;
-    total_word_count += tmp_spvbinary.wordCount;
+    i = i.Erase();
   }
 
-  // id_upper_bound is currently equal to the highest id being used, so add
-  // one to get the strict upper bound
-  ++id_upper_bound;
-
-  const bool create_library = (options & SPV_LINKER_OPTION_CREATE_LIBRARY) != 0u;
-
-  *linked_binary = new spv_binary_t;
-  (*linked_binary)->code = new uint32_t[total_word_count];
-
-  ret = MergeModules(hijack_context, tmp_binaries, num_binaries,
-                            id_upper_bound, max_version, **linked_binary,
-                            create_library);
-  if (ret != SPV_SUCCESS) {
-    spvBinaryDestroy(*linked_binary);
-    return ret;
+  // Remove import linkage attributes
+  for (auto i = linkedModule->annotation_begin(); i != linkedModule->annotation_end();) {
+    if (i->opcode() != SpvOpDecorate ||
+        static_cast<spv::Decoration>(i->GetSingleWordOperand(1u)) != spv::Decoration::LinkageAttributes ||
+        static_cast<spv::LinkageType>(i->GetSingleWordOperand(3u)) != spv::LinkageType::Import) {
+      ++i;
+      continue;
+    }
+    i = i.Erase();
   }
 
-  return ret;
-}
-
-} // anonymous namespace
-
-namespace spvtools {
-
-// Structs for holding the data members for SpvTools.
-struct SpirvLinker::Impl {
-  explicit Impl(spv_target_env env) : context(spvContextCreate(env)) {
-    // The default consumer in spv_context_t is a null consumer, which provides
-    // equivalent functionality (from the user's perspective) as a real consumer
-    // does nothing.
+  // Remove export linkage attributes
+  if (!(options & SPV_LINKER_OPTION_CREATE_LIBRARY)) {
+    for (auto i = linkedModule->annotation_begin(); i != linkedModule->annotation_end();) {
+      if (i->opcode() != SpvOpDecorate ||
+          static_cast<spv::Decoration>(i->GetSingleWordOperand(1u)) != spv::Decoration::LinkageAttributes ||
+          static_cast<spv::LinkageType>(i->GetSingleWordOperand(3u)) != spv::LinkageType::Export) {
+        ++i;
+        continue;
+      }
+      i = i.Erase();
+    }
   }
-  ~Impl() { spvContextDestroy(context); }
 
-  spv_context context;  // C interface context object.
-};
-
-SpirvLinker::SpirvLinker(spv_target_env env) : impl_(new Impl(env)) {}
+  // TODO recompact IDs
 
-SpirvLinker::~SpirvLinker() {}
+  ModuleHeader header;
+  header.magic_number = spv::MagicNumber;
+  header.version = spv::Version; // FIXME
```

Comment:
Please use TODO(your-name) rather than FIXME.
Also, what needs fixing here?  Say you've got a decision to make.  Use the max version of any module you've seen?  I think that's probably right.